src/ceph-volume/ceph_volume/devices/lvm/batch.py

```diff
@@ -316,79 +387,216 @@ def main(self):
+            exit(0)
+        if fast_devices and not len(fast_allocations) == num_osds:
+            mlogger.error('{} fast allocations != {} num_osds'.format(
+                len(fast_allocations), num_osds))
+            exit(1)
+
+        very_fast_allocations = self.fast_allocations(very_fast_devices,
+                                                      requested_osds,
+                                                      num_osds,
+                                                      'block_wal')
+        if very_fast_devices and not very_fast_allocations:
+            mlogger.info('{} very fast devices were passed, but none are available'.format(len(very_fast_devices)))
+            exit(0)
+        if very_fast_devices and not len(very_fast_allocations) == num_osds:
+            mlogger.error('{} very fast allocations != {} num_osds'.format(
+                len(very_fast_allocations), num_osds))
+            exit(1)
+
+        for osd in plan:
+            if fast_devices:
+                osd.add_fast_device(*fast_allocations.pop(),
+                                    type_=fast_type)
+            if very_fast_devices and args.bluestore:
+                osd.add_very_fast_device(*very_fast_allocations.pop(),
+                                        type_='block.wal')
+        return plan
+
+    def fast_allocations(self, devices, requested_osds, new_osds, type_):
+        ret = []
+        if not devices:
+            return ret
+        phys_devs = [d for d in devices if d.is_device]
+        lvm_devs = [d.lvs[0] for d in list(set(devices) -
+                                           set(phys_devs))]
+        mlogger.debug(('passed {} devices: {} physical,'
+                       ' {} LVM').format(type_, len(phys_devs), len(lvm_devs)))
+
+        ret.extend(get_lvm_fast_allocs(lvm_devs))
+
+        # fill up uneven distributions across fast devices: 5 osds and 2 fast
+        # devices? create 3 slots on each device rather then deploying
+        # heterogeneous osds
+        if (requested_osds - len(lvm_devs)) % len(phys_devs):
+            fast_slots_per_device = int((requested_osds - len(lvm_devs)) / len(phys_devs)) + 1
+        else:
+            fast_slots_per_device = int((requested_osds - len(lvm_devs)) / len(phys_devs))
+
+
+        ret.extend(get_physical_fast_allocs(phys_devs,
+                                            type_,
+                                            fast_slots_per_device,
+                                            new_osds,
+                                            self.args))
+        return ret
+
+    class OSD(object):
+        '''
+        This class simply stores info about to-be-deployed OSDs and provides an
+        easy way to retrieve the necessary create arguments.
+        '''
+        def __init__(self, data_path, rel_size, abs_size, slots, id_):
+            self.id_ = id_
+            self.data = (data_path, rel_size, abs_size, slots)
```

Comment:
What about using namedtuples instead of regular tuples. I guess this will make it way easier to comprehend in the future.

``` python

from typing import NamedTuple
# alternatively 
# from collections import namedtuple
# if backporting to a python2 based platform is required.

class DeviceDescription(NamedTuple):
    path: str 
    rel_size: int 
    abs_size: int
    slots: int
    type_: str = None
    
self.data = DeviceDescription(path=data_path,
                              rel_size=data_size,
                              abs_size=abs_size,
                              slots=slots
                              )

```
This avoids all those awkward calls like `self.data[0] and self.data[2]` in favor of the `.` notation.

Reply:
Yeah good point. Thought I can get away with this since its only used within the OSD class, but I'm probably just lazy :)
Coming up.